pkg/tfbridge/validate_input_types.go

```diff
@@ -35,215 +35,249 @@ func NewInputValidator(urn resource.URN, schema pschema.PackageSpec) *PulumiInpu
+		if failure != nil {
+			failures = append(failures, failure...)
 		}
 	}
-	return complexSpec, typeSpec, requiredProps
+	return failures
 }
 
-// validateTypeSpec is the main function for validating a PropertyValue against the pulumi schema.
-// It is a recursive function that will validate nested types and arrays. It will return a list of
-// type failures if any are found
-func (v *PulumiInputValidator) validateTypeSpec(
-	propertyKey string,
+// validatePropertyValue is the main function for validating a PropertyValue against the Pulumi Schema. It is a
+// recursive function that will validate nested types and arrays. Returns a list of type failures if any are found.
+func (v *PulumiInputValidator) validatePropertyValue(
 	propertyValue resource.PropertyValue,
 	typeSpec pschema.TypeSpec,
 	propertyPath resource.PropertyPath,
-	requiredProps []string,
 ) []TypeFailure {
-	// for now don't validate discriminators
-	if typeSpec.Discriminator != nil {
-		return nil
-	}
 	// don't type check
-	// - resource references
-	// - assets
-	// - archives
-	// - computed values
+	// - resource references (not yet)
+	// - assets (not yet)
+	// - archives (not yet)
+	// - computed values (they are allowed for any type)
+	// - unknown output values (similar to computed)
+	// - nulls (they are allowed for any type, and for missing required properties see validatePropertyMap)
 	if propertyValue.IsResourceReference() ||
 		propertyValue.IsAsset() ||
 		propertyValue.IsArchive() ||
-		propertyValue.IsComputed() {
+		propertyValue.IsComputed() ||
+		propertyValue.IsNull() ||
+		(propertyValue.IsOutput() && !propertyValue.OutputValue().Known) {
 		return nil
 	}
 
-	// handle non-required properties that are null
-	if propertyValue.IsNull() {
-		if slices.Contains(requiredProps, propertyKey) {
-			return []TypeFailure{
-				{
-					ResourcePath: propertyPath.String(),
-					Reason:       fmt.Sprintf("property %s is required", propertyKey),
-				},
-			}
-		}
+	// for known first-class outputs simply validate their known value
+	if propertyValue.IsOutput() && propertyValue.OutputValue().Known {
+		elementValue := propertyValue.OutputValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
 	}
 
-	failures := []TypeFailure{}
-	if propertyKey != "" {
-		propertyPath = append(propertyPath, propertyKey)
+	// for secrets validate their inner value
+	if propertyValue.IsSecret() {
+		elementValue := propertyValue.SecretValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
 	}
 
-	// perform some initial easy type matching. fail fast.
-	possible, ok := v.matchType(propertyValue, typeSpec)
-	if !ok {
-		return []TypeFailure{
-			{
+	// now we are going to switch on the desired type
+	//
+	// a good reference here for the semantics of the TypeSpec is
+	// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L881
```

Comment:
This is the key reference for me to read to copy how it does case analysis to untangle the representation of TypeSpec. Including here in a comment.